pennylane/qnodes/base.py

```diff
@@ -414,70 +442,45 @@ def qfunc(a, w):
 
         Args:
             args (tuple[Any]): Positional arguments passed to the quantum function.
-                During the construction we are not concerned with the numerical values, but with
-                the nesting structure.
-                Each positional argument is replaced with a :class:`~.variable.VariableRef` instance.
             kwargs (dict[str, Any]): Auxiliary arguments passed to the quantum function.
+                They are only represented with VariableRefs if the circuit is immutable.
+
+        Returns:
+            nested list[VariableRef], Union[nested list[VariableRef], None]: references for args, references for kwargs
         """
-        # Get the name of the qfunc's arguments
-        full_argspec = inspect.getfullargspec(self.func)
-
-        # args
-        variable_name_strings = []
-        for variable_name, variable_value in zip(full_argspec.args, args):
-            variable_name_strings.append(
-                self._determine_structured_variable_name(variable_value, variable_name)
+        # positional args
+        variable_names = []
+        # the positional args must come first in the signature
+        for prefix, a in zip(self.func.sig, args[: self.func.n_pos_nd]):
+            variable_names.append(self._determine_structured_variable_name(a, prefix))
+
+        # var-positional args come next
+        n_var_pos = len(args) - self.func.n_pos_nd
+        if n_var_pos > 0:
+            variable_names.append(
+                self._determine_structured_variable_name(args[-n_var_pos:], self.func.var_pos)
             )
 
-        # varargs
```

Comment:
I didn't actually know that varargs were forbidden

Reply:
Currently they are allowed, technically, since the are positional args (and I believe one test actually uses this feature). But maybe they should be forbidden.

Reply:
As you've probably noticed, I've been expanding the BaseQNode docstring quite a bit to try and explicitly list the assumptions the code makes about what is allowed in terms of the qfunc signature, calling convention and mutability of the circuit. I think it'd be a good idea to have a thorough discussion about this and codify it in an ADR.

Reply:
The ADR is here: https://github.com/XanaduAI/pennylane-adrs/pull/2